### PR TITLE
Path parameterization and seven segment profile

### PR DESCRIPTION
**Pull request by jonbinney**
Example of a 2d joint-space path getting parameterized according to a single variable, from `rosrun traj parameterize_path_example`.
parameterize_path

Example of a seven segment 1-d motion profile (starting and ending with a=0, v=0) from `rosrun traj seven_segment_example`

seven_segment

### CODE REVIEW

**Comment by bhomberg** [comment]
@jonbinney this looks awesome!  Do you want a review?

**Comment by gavanderhoorn** [comment]
@mahmoud-a-ali

**Comment by jonbinney** [comment]
@bhomberg yes, a review would be great, thanks!

**Comment by mahmoud-a-ali** [comment]
Great results, the single variable idea is really interesting for the synchronization part

**Comment by jonbinney** [comment]
@mahmoud-a-ali to be clear - this isn't the full solution yet. The path parameterization and the seven segment motion profile generation are separate in this PR. So the very pretty graph of the jerk/acc/vel/pos is just for a single segment 1dof trajectory. I'll combine them in my next PR to to handle multi-dof and multi-segment.

**Comment by mahmoud-a-ali** [comment]
I saw that it is one segment 1dof, I meant that the s variable could be interesting for the syncronnization part of multiple dof!

**Comment by gavanderhoorn** [comment]
Note that we could already test single dof trajectories on real robots.

We'd just keep the other joints at their current angles.

**Comment by jonbinney** [comment]
@bhomberg bump - could you review?

File: traj/src/traj/plot.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    if a_max is not None:
+        axes[2].plot(plot_times, [a_max] * len(plot_times), '--', color='red')
+        axes[2].plot(plot_times, [-a_max] * len(plot_times), '--', color='red')
+    axes[3].plot(plot_times, jerks)
+    axes[3].set_ylabel('jerk')
+    if j_max is not None:
+        axes[3].plot(plot_times, [j_max] * len(plot_times), '--', color='red')
+        axes[3].plot(plot_times, [-j_max] * len(plot_times), '--', color='red')
+
+def plot_2d_path(piecewise_function, npoints, label='path points'):
```
Comment (bhomberg) [inline traj/src/traj/plot.py]: Can you expand on how you're handling fig/axes here?  You don't call plt.show() directly, leaving that to the caller, but you also don't take in fig/axes as arguments, not giving the caller full control over that.  Not sure if this is a standard way to handle matplotlib functions, but it seems like it could cause some unexpected behavior depending on how else the user is using matplotlib.

File: traj/src/traj/plot.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    """
+    Plot a 2d path that is represented as a piecewise function of a single variable.
+    """
+    S, path_points = piecewise_function.sample(npoints)
+    plt.plot(path_points[:,0], path_points[:,1], 'r.-', label=label)
```
Comment (bhomberg) [inline traj/src/traj/plot.py]: Might be worth using a different color than red, since above you're using red as boundary markers.

File: traj/src/traj/parameterize_path.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+def parameterize_path(path):
+    """
+    Represent the given joint-space path as a function q = f(s).
+
+    To make this useful as a first step to creating smooth trajectories, we consruct the
+    parameterization such that it has a few important properties:
+
+    1. s = 0 is the start point of the path in joint space.
+    2. As s increases, we move along the path monononically.
```
Comment (bhomberg) [inline traj/src/traj/parameterize_path.py:13]: monotonically

File: traj/src/traj/parameterize_path.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    3. The length of path traversed for a given increase in s is constant.
+
+    Out of convenience, we choose a parameterization such that the value of s is equal
+    to the length of path travesed. This "length" is in N-dimensional joint space, but
+    otherwise is the same as we would compute a path length.
+
+    Put another way, the path length from s=0 to s=s' is equal to the integral from 0 to s'
```
Comment (bhomberg) [inline traj/src/traj/parameterize_path.py]: Suggestion: use S instead of s (or some other distinction) when talking about S, the length of the path (a constant for a given path) vs s, the variable along which we will have parameterized the function.  I think it would make this section slightly clearer.

File: traj/src/traj/parameterize_path.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    of the norm of the derivative of the parameterized path function w.r.t. the variable s
+    """
+    s = Symbol('s')
+    boundaries = [0.0]
+    functions = []
+    for q0, q1 in zip(path[:-1], path[1:]):
```
Comment (bhomberg) [inline traj/src/traj/parameterize_path.py:29]: Can you add some comments here about what this is doing, math-wise?  What are q0 and q1?  What are boundaries?

File: traj/src/traj/piecewise_function.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import numpy as np
+from sympy import Matrix, Piecewise, Symbol
+
+class PiecewiseFunction:
+    """
+    A piecewise function of a single variable.
+
+    Uses sympy as its function representation, but because we know we are a function of
```
Comment (bhomberg) [inline traj/src/traj/piecewise_function.py:8]: Request for more explanation about how you're representing piecewise functions

File: traj/src/traj/piecewise_function.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    one variable, we are able to compute integrals correctly (unlike sympy's builtin piecewise
+    function).
+    """
+    def __init__(self, boundaries, functions, independent_variable):
+        self.boundaries = boundaries
+        self.functions = functions
+        self.independent_variable = independent_variable
+        assert len(boundaries) - 1 == len(functions)
+
+    def __call__(self, value):
```
Comment (bhomberg) [inline traj/src/traj/piecewise_function.py:18]: Maybe clarify this variable as input or input_value to be clearer.

Comment (bhomberg) [reply traj/src/traj/piecewise_function.py:18]: Also, what happens if value is below 0 or greater than the max bound covered?  Shouldn't it throw a clear error?

File: traj/src/traj/seven_segment.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
+from sympy import integrate, Symbol
+from sympy.core.numbers import Float
+
+from .piecewise_function import PiecewiseFunction
+
+def fit_seven_segment(p_start, p_end, v_max, a_max, j_max):
```
Comment (bhomberg) [inline traj/src/traj/seven_segment.py:6]: Maybe rename to clarify that it will stop at both ends, since I assume we will later have one that takes v_start, v_end, etc and doesn't assume that.  Maybe fit_seven_segment_stopped_at_ends?

File: traj/src/traj/seven_segment.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        a_max = T_jmax * j_max
+
+    # Compute the minimum distance that each case can travel. D_thr1 is the minimum distance for a
+    # trajectory that hits both max acceleration and max velocity. D_thr2 is the minimum distance
+    # for a trajectory that hits max acceleration but not max velocity.
+    D_thr1 = (a_max * v_max) / j_max + v_max**2 / a_max
+    D_thr2 = 2.0 * a_max**3 / j_max**2
+
+    D = p_end - p_start
+    if D >= D_thr1:
```
Comment (bhomberg) [inline traj/src/traj/seven_segment.py:44]: What if abs(D) >= D_thr1, but p_start > p_end?  We'd hit both -v_max and -a_max.

File: traj/src/traj/seven_segment.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        # We hit both v_max and a_max
+        T_j = T_jmax
+        T_a = T_amax
+        T_v = (D - D_thr1) / v_max
+    elif D > D_thr2:
```
Comment (bhomberg) [inline traj/src/traj/seven_segment.py:49]: Ditto for negative case.

File: traj/src/traj/seven_segment.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    times = [0.0]
+    jerk_functions = []
+    acceleration_functions = []
+    velocity_functions = []
+    position_functions = []
+    t = Symbol('t')
+    for j0, T in segment_jerks_and_durations:
+        times.append(times[-1] + T)
+        j = Float(j0)
+        a = integrate(j, t) + a0
```
Comment (bhomberg) [inline traj/src/traj/seven_segment.py:76]: Add comment that these are integrated from the beginning, not just for this segment (otherwise you'd need a[-1], not a0).

Comment (bhomberg) [reply traj/src/traj/seven_segment.py:76]: Actually, this sounds more like we're calculating delta a / delta v / delta p functions, not pure a / v/ p functions -- shouldn't we not add a0 / etc and offset the da / dv / dp by the previous a / v/ p value instead?

File: traj/test/test_seven_segment.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import nose
+import numpy as np
+
+from traj import fit_seven_segment
+
+def check_fit_seven_segment(p_start, p_end, v_max, a_max, j_max):
```
Comment (bhomberg) [inline traj/test/test_seven_segment.py:6]: Consider adding tests that confirm that max values are never violated.

Consider adding tests for start position is higher than end position (negative displacement).

File: exploration_scripts/seven_segment_trajectory [hunk 13]
```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env python
```
Comment (bhomberg) [inline exploration_scripts/seven_segment_trajectory]: This file seems to have overlapping code with many other files -- was it a starter file that was intended to be replaced by better organized files?

**Review by bhomberg — CHANGES_REQUESTED** [review]

File: traj/src/traj/plot.py (hunk 1 above)
Comment (jonbinney) [reply traj/src/traj/plot.py]: Yeah, i was being lazy. Updated so that the caller passes in the axes they want the funciton to use.

File: traj/src/traj/parameterize_path.py (hunk 4 above)
Comment (jonbinney) [reply traj/src/traj/parameterize_path.py]: Done

File: traj/src/traj/parameterize_path.py (hunk 5 above)
Comment (jonbinney) [reply traj/src/traj/parameterize_path.py:29]: Added explanation comment.

File: traj/src/traj/seven_segment.py (hunk 9 above)
Comment (jonbinney) [reply traj/src/traj/seven_segment.py:44]: Hmmm.... I'll have to think about this. I think I'm going to go ahead and merge, knowing that this function doesn't handle all cases. Lets add a fix and a unit test for this case in a future PR.

File: traj/src/traj/seven_segment.py (hunk 11 above)
Comment (jonbinney) [reply traj/src/traj/seven_segment.py:76]: They're integrated over time from the beginning, as you said in your first comment. Added a comment to make this clearer.

File: traj/test/test_seven_segment.py (hunk 12 above)
Comment (jonbinney) [reply traj/test/test_seven_segment.py:6]: Good point - I'll add these in another PR. Issue created here: https://github.com/iron-ox/trajectory_smoothing/issues/5

File: exploration_scripts/seven_segment_trajectory (hunk 13 above)
Comment (jonbinney) [reply exploration_scripts/seven_segment_trajectory]: Oops! Yes. Deleted.
